### ISSUE TITLE
change propagator updates parent too often

### ISSUE THREAD

**Issue by schiesbn** [owncloud/core#18359]
I discovered a issue with sharing in combination with encryption. If you share a folder containing 6 files with 4 users the share operation will take 24 seconds and runs 2500 queries !!!!

The problem seems to be that every time we write a file- or share-key the cache propagator propagates the changes to all parents although the cache updater already calls the propagator for each parent.

See the profile below.
https://blackfire.io/profiles/03cdaffa-bc31-4f17-921e-df0da04ed0c3/graph

@icewind1991 This needs urgent fixing, can you have a look? Thanks!

cc @nickvergessen @DeepDiver1975

**Comment by icewind1991** [comment]
> although the cache updater already calls the propagator for each parent.

That's not the case, there are 2 propagation steps, correcting folder sizes and propagating mtimes/etags, they are done separate because they have different requirements

**Comment by icewind1991** [comment]
The problem is that the encryption code makes 48 filessystem calls with an average folder depth of 8.

Since there is no knowledge in the filesystem layer about how things are grouped the only way to handle that is to do 48 * 8 * 2 = 786 cache updates (the remaining queries are cache reads from what I can tell)

**Comment by schiesbn** [comment]
> The problem is that the encryption code makes 48 filessystem calls with an average folder depth of 8.

We still need to find a solution how to improve this. On the encryption level there is not much we can do about it. We need to write a file-key for each file and a share key for each user and create the folder before we can create the files. I don't see how this can be improved on the encryption side, we perform the absolute minimum of operations in order to update the keys.

I also don't really understand how the 786 cache updated leads to 2578 queries at the end. According from the graph they come all from OC\File\Cache\Cache::update, something which shouldn't be triggered by a read operation

**Comment by schiesbn** [comment]
> That's not the case, there are 2 propagation steps, correcting folder sizes and propagating mtimes/etags, they are done separate because they have different requirements

Just thinking: Couldn't this for example be done in a way that we first collect the data for mtime/etag and folder size in two steps but then perform the actual db update in one step once we have collected all the new values?

**Comment by icewind1991** [comment]
> According to the graph they all come from OC\File\Cache\Cache::update

The graph is lying, if you click on `executeAudited` you see that it has 6 callers, the graph only shows callers which  are significant in time and since updates are way more expensive (~87% for 786 updates, and ~5% for the remaining ~1800 reads)

**Comment by icewind1991** [comment]
> We still need to find a solution how to improve this. On the encryption level there is not much we can do about it. We need to write a file-key for each file and a share key for each user and create the folder before we can create the files. I don't see how this can be improved on the encryption side, we perform the absolute minimum of operations in order to update the keys.

Currently the only way to improve it is to bypass the view, work directly on the storage and call the updater yourself when you're done doing all operations. Since the crypt code knows how it will be using the filesystem it doesn't need to keep the cache uptodate during every step, just needs to make sure it's consistent at the end.

I've been thinking about adding some sort of "operation grouping" api to the filesystem to make things like this easier, something like

```php
$view->grouped(function($view){
    $view->doSomething();
    $view->doMore();
});

Where the cache updates will only happen once the entire group is done

**Comment by schiesbn** [comment]
> The graph is lying, if you click on executeAudited you see that it has 6 callers, the graph only shows callers which are significant in time and since updates are way more expensive (~87% for 786 updates, and ~5% for the remaining ~1800 reads)

So it seems that we still need to optimize the ~95%, even if it is "only" 786 queries from ~2500
If we could update each file cache entry only once instead of two we could already reduce the time almost by 50%. It would be still way to much but a step in the right direction.

I also checked the encryption code once more and I don't see where the read opertions come from, we don't do any of them while writing the keys. Maybe some of them come from the update process and probably some from the sharing code? But it is probably not worth thinking hard about optimizing 1,9% of this issue.

> I've been thinking about adding some sort of "operation grouping" api to the filesystem to make things like this easier, something like...

could be a step in the right direction. Any chance we can see something like this for 8.2?

**Comment by PVince81** [comment]
Is this about etag / size propagation when writing to "files_encryption" or to "files" ?
Because if this is happening outside of "files" we can probably skip such updates as we don't need etags there.

**Comment by schiesbn** [comment]
It's all updates, size, etag and mtime. According to @icewind1991 every file/folder is touched twice. Once for to update the size and once to update the mtime and etag

**Comment by schiesbn** [comment]
@michaelstingl for your information

**Comment by PVince81** [comment]
@icewind1991 @schiesbn any update ? Is there a chance for a short term fix for this ?

**Comment by icewind1991** [comment]
> Is there a chance for a short term fix for this ?

Dont think so

**Comment by icewind1991** [comment]
@DeepDiver1975 do we want to keep this scheduled for 8.2?

**Comment by DeepDiver1975** [comment]
@cmonteroluque given @icewind1991's judgment we better move this

**Comment by karlitschek** [comment]
Hmm. I have to agree. Let's move it

**Comment by cmonteroluque** [comment]
@DeepDiver1975 @karlitschek it's brutal performance... @icewind1991 forget about a quick fix, do we have a fix that we could apply to 9.0 or at this point we're not even sure? If nothing else, we should have this top of mind as part of 9.0 planning

**Comment by icewind1991** [comment]
Yes, there is a solution we can use for 9.0

**Comment by DeepDiver1975** [comment]
> Yes, there is a solution we can use for 9.0

@cmonteroluque move? THX

**Comment by cmonteroluque** [comment]
@DeepDiver1975 ok, moving...

**Comment by karlitschek** [comment]
@schiesbn We shouldn't forget this. I suggest to work on that as soon as master is open for commits again.

**Comment by MorrisJobke** [comment]
Maybe #10078 also gives more clues, what is going on.

**Comment by PVince81** [comment]
@icewind1991 @schiesbn any progress on this ?
@icewind1991 I saw your recent PRs about etag propagation, not sure if related ?

**Comment by MorrisJobke** [comment]
@icewind1991 What is the status here? The change propagator was removed #20875 and now the per storage propagation is in place. Is this still an issue here?

**Comment by icewind1991** [comment]
#20875 improves the situation a bit but the issue still exists

**Comment by PVince81** [comment]
Any update ?

**Comment by icewind1991** [comment]
9.0 will do things better but there is still room for improvements.

iirc 8.2 does about 2N + 2M +4 queries (N being the folder depth within the storage, M being the total folder depth)

master currently does 2N+4 queries per write for updating the cache

I think I can get it down to 4 queries at least

**Comment by danielkoch** [comment]
Will this fix also be in the next 8.2 version?

**Comment by PVince81** [comment]
It depends on the fix, its extent is not clear yet.

**Comment by icewind1991** [comment]
> Will this fix also be in the next 8.2 version?

No, the fix relies on some of the larger changes that were made in the cache area for 9.0

**Comment by PVince81** [comment]
@icewind1991 the original report mentions propagating etag for encryption keys. Would it be possible to exclude this from propagation and limit propagation to "files/" and external storage ? (not sure if this would have other consequences).

**Comment by icewind1991** [comment]
Should this still be a sev-1 for 9.0 with the improvements that are already in 9.0?

**Comment by PVince81** [comment]
@icewind1991 good point, @schiesbn's original ticket was discovered in 8.2.2.

Would still be good to have a backportable quickfix where etags/mtimes are not propagated for encryption keys, as it makes no sense.

**Comment by icewind1991** [comment]
I'm not sure about the impact of such a change, code might rely on things like sizes being updated correctly

**Comment by PVince81** [comment]
@icewind1991 I don't think the encryption code requires the size of the key folders, also that is not counted in the quota.

@schiesbn what do you think ?

**Comment by icewind1991** [comment]
iirc trashbin uses the size

**Comment by PVince81** [comment]
To limit the impact, I suggest limiting the exclusion to "files_encryption"

**Comment by MorrisJobke** [comment]
Not really a blue ticket anymore. Case was closed.

**Comment by icewind1991** [comment]
With #22733 there is a ~24% improvement in run time between 8.2 and 9.0 ([comparison](https://blackfire.io/profiles/compare/aa6e8b83-75a0-48fa-aaac-1f121ae7d644/graph))

**Comment by icewind1991** [comment]
With #22733 and #22648 combined ~34% ([comparison](https://blackfire.io/profiles/compare/00e1ffe0-0be2-475b-93ea-290feac75447/graph))

**Comment by icewind1991** [comment]
note that these numbers come from a setup with a local sqlite db stored on a ramdisk, the impact of the saving is likely to be larger in a "real" setup

**Comment by icewind1991** [comment]
> To limit the impact, I suggest limiting the exclusion to "files_encryption"

#22736 for ~40% speedup compared to master ([comparison](https://blackfire.io/profiles/compare/bb5c8329-4e77-4694-8f5c-3437dff43cb5/graph))

**Comment by PVince81** [comment]
@icewind1991 is there any improvement already without https://github.com/owncloud/core/pull/22733 ? At least with https://github.com/owncloud/core/pull/22648 alone ?

**Comment by PVince81** [comment]
I don't want to merge https://github.com/owncloud/core/pull/22733 now, it's too risky (moved to 9.0.1).

https://github.com/owncloud/core/pull/22648 seems to be an oversight from another perf fix you did before, so we need to get that one in too for 9.0 to get some benefit. Is that correct ?

**Comment by icewind1991** [comment]
(all comparisons are made against stable8.2)

> @icewind1991 is there any improvement already without #22733 ? At least with #22648 alone ?

Just #22648 [~19%](https://blackfire.io/profiles/compare/e23664f7-bf4e-4c7d-9aff-27a31bd0c339/graph)

**Comment by icewind1991** [comment]
> I don't want to merge #22733 now, it's too risky (moved to 9.0.1).

> #22648 seems to be an oversight from another perf fix you did before, so we need to get that one in too for 9.0 to get some benefit. Is that correct ?

I think you're mixing up the 2, #22733 is a fairly small fix that I missed before, #22648 I agree is to risky

**Comment by PVince81** [comment]
Yeah, mixed up sorry.

**Comment by MorrisJobke** [comment]
* [ ] #22648
* [ ] #22733 
* [ ] #22736

**Comment by MorrisJobke** [comment]
Only #22648 is not merged and will not get into 9.1 -> moving this issue to 9.1 too, because the 9.0 issues are resolved.

**Comment by MorrisJobke** [comment]
@PVince81 Or should we create a new issue?

**Comment by PVince81** [comment]
I'm fine keeping this issue.

**Comment by icewind1991** [comment]
Can we close this issue with #22648 merged
